### PR TITLE
added invocation logs

### PR DESCRIPTION
**Pull request by KavishShah09**

### CODE REVIEW

File: gateway/modules/eventing/http.go [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+// MakeInvocationHTTPRequest fires an http request and returns a response
+func (s *Module) MakeInvocationHTTPRequest(ctx context.Context, method, url, token, scToken string, params, vPtr interface{}) error {
+	// Marshal json into byte array
+	data, _ := json.Marshal(params)
+
+	// Make a request object
+	req, err := http.NewRequestWithContext(ctx, method, url, bytes.NewBuffer(data))
+	payload, er := interfaceToByteArray(params)
+	if er != nil {
+		invocationDoc := &model.InvocationDocument{
```
Comment (YourTechBud) [inline gateway/modules/eventing/http.go]: You didn't  set request body  and event id

File: gateway/modules/eventing/http.go [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+}
+
+// MakeInvocationHTTPRequest fires an http request and returns a response
+func (s *Module) MakeInvocationHTTPRequest(ctx context.Context, method, url, token, scToken string, params, vPtr interface{}) error {
+	// Marshal json into byte array
+	data, _ := json.Marshal(params)
+
+	// Make a request object
+	req, err := http.NewRequestWithContext(ctx, method, url, bytes.NewBuffer(data))
+	payload, er := interfaceToByteArray(params)
```
Comment (YourTechBud) [inline gateway/modules/eventing/http.go]: data is the payload

File: gateway/modules/eventing/http.go [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+	if er != nil {
+		invocationDoc := &model.InvocationDocument{
+			InvocationTime:     time.Now().String(),
```
Comment (YourTechBud) [inline gateway/modules/eventing/http.go]: use RFC3339 format

File: gateway/modules/eventing/http.go [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+			ResponseStatusCode: 0,
+			ErrorMessage:       er.Error(),
+		}
+		createRequest := &model.CreateRequest{Document: invocationDoc, Operation: utils.All, IsBatch: true}
```
Comment (YourTechBud) [inline gateway/modules/eventing/http.go]: Operation should be one

File: gateway/modules/eventing/http.go [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+		if err := s.crud.InternalCreate(ctx, s.config.DBType, s.project, invocationLogs, createRequest, false); err != nil {
+			return errors.New("eventing module couldn't log the request - " + err.Error())
+		}
+		return er
+	}
+	if err != nil {
+		invocationDoc := &model.InvocationDocument{
```
Comment (YourTechBud) [inline gateway/modules/eventing/http.go]: You didnt set thhe event id

File: gateway/modules/eventing/http.go [hunk 6]
```diff
@@ -0,0 +1,141 @@
...
+
+	// if s.storeType && s.isConsulConnectEnabled && strings.Contains(url, "https") && strings.Contains(url, ".consul") {
+	// 	 client = s.consulService.HTTPClient()
+	// }
+
+	req = req.WithContext(ctx)
+	resp, err := client.Do(req)
+	responseBody, e := ioutil.ReadAll(resp.Body)
+	if e != nil {
+		invocationDoc := &model.InvocationDocument{
```
Comment (YourTechBud) [inline gateway/modules/eventing/http.go]: set the event id

File: gateway/modules/eventing/http.go [hunk 7]
```diff
@@ -0,0 +1,141 @@
...
+			ResponseBody:       string(responseBody),
+			ErrorMessage:       err.Error(),
+		}
+		createRequest := &model.CreateRequest{Document: invocationDoc, Operation: utils.All, IsBatch: true}
+		if err := s.crud.InternalCreate(ctx, s.config.DBType, s.project, invocationLogs, createRequest, false); err != nil {
+			return errors.New("eventing module couldn't log the request - " + err.Error())
+		}
+		return err
+	}
+	defer utils.CloseTheCloser(resp.Body)
```
Comment (YourTechBud) [inline gateway/modules/eventing/http.go]: Do this right after you get the response

File: gateway/modules/eventing/http.go [hunk 8]
```diff
@@ -0,0 +1,141 @@
...
+
+	if err := json.NewDecoder(resp.Body).Decode(vPtr); err != nil {
```
Comment (YourTechBud) [inline gateway/modules/eventing/http.go]: Use simple json marshal

File: gateway/modules/eventing/http.go [hunk 9]
```diff
@@ -0,0 +1,141 @@
...
+		invocationDoc := &model.InvocationDocument{
```
Comment (YourTechBud) [inline gateway/modules/eventing/http.go]: set the event id

File: gateway/modules/eventing/http.go [hunk 10]
```diff
@@ -0,0 +1,141 @@
...
+		}
+		createRequest := &model.CreateRequest{Document: invocationDoc, Operation: utils.All, IsBatch: true}
+		if err := s.crud.InternalCreate(ctx, s.config.DBType, s.project, invocationLogs, createRequest, false); err != nil {
+			return errors.New("eventing module couldn't log the request - " + err.Error())
+		}
+		return err
+	}
+
+	if resp.StatusCode < 200 || resp.StatusCode >= 300 {
+		invocationDoc := &model.InvocationDocument{
```
Comment (YourTechBud) [inline gateway/modules/eventing/http.go]: You didn't set the event id

File: gateway/modules/eventing/http.go (hunk 8 above)
Comment (KavishShah09) [reply gateway/modules/eventing/http.go]: Do you mean unmarshal?

**Comment by YourTechBud** [comment]
This fixes #807

File: gateway/modules/eventing/http.go [hunk 11]
```diff
@@ -0,0 +1,110 @@
...
+	if err != nil {
+		if err := s.logInvocation(ctx, eventDoc.ID, data, 0, "", err.Error()); err != nil {
+			logrus.Errorf("eventing module couldn't log the invocation - %s", err.Error())
+			return err
+		}
+		return err
+	}
+
+	if err := json.Unmarshal(responseBody, vPtr); err != nil {
+		err := s.logInvocation(ctx, eventDoc.ID, data, resp.StatusCode, string(responseBody), err.Error())
```
Comment (YourTechBud) [inline gateway/modules/eventing/http.go]: Handle this error inline

File: gateway/modules/eventing/http.go [hunk 12]
```diff
@@ -0,0 +1,110 @@
...
+		if err != nil {
+			logrus.Errorf("eventing module couldn't log the invocation - %s", err.Error())
+			return err
+		}
+		return err
+	}
+
+	if resp.StatusCode < 200 || resp.StatusCode >= 300 {
+		err := s.logInvocation(ctx, eventDoc.ID, data, resp.StatusCode, string(responseBody), err.Error())
```
Comment (YourTechBud) [inline gateway/modules/eventing/http.go]: Handle this error inline

File: gateway/modules/eventing/http.go [hunk 13]
```diff
@@ -0,0 +1,110 @@
...
+		if err != nil {
+			logrus.Errorf("eventing module couldn't log the invocation - %s", err.Error())
+			return err
+		}
+		return errors.New("service responded with status code " + strconv.Itoa(resp.StatusCode))
```
Comment (YourTechBud) [inline gateway/modules/eventing/http.go]: Use fmt.Errorf here

**Pull request merged**